qiime2/core/type/signature.py

```diff
@@ -289,16 +289,36 @@ def decode_parameters(self, **kwargs):
 
     def check_types(self, **kwargs):
         for name, spec in self.signature_order.items():
-            if kwargs[name] not in spec.qiime_type:
-                # A type mismatch is unacceptable unless the value is None
-                # and this parameter's default value is None.
-                if not (spec.has_default() and
-                        spec.default is None and
-                        kwargs[name] is None):
+            # A type mismatch is unacceptable unless the value is None
+            # and this parameter's default value is None.
+            if ((kwargs[name] not in spec.qiime_type) and
+                    not (spec.has_default() and spec.default is None
+                         and kwargs[name] is None)):
+
+                # handle visualization passed as parameter
```

Comment:
Do you think this comment adds anything to the discussion? My concern is that it basically echoes the line below. My rule of thumb: code declares _how_, comments declare _why_.

One way to make the code "speak" more clearly is to "rename" `kwargs[name]` somewhere up here:

```python
parameter = kwargs[name]

if isinstance(parameter, qiime2.sdk.Visualization):
    ...
```

Now I read that line as "if the parameter is an instance of `Visualization`..."

Reply:
Love this. Thanks!